File: himsha-sdk-dart/lib/src/programs/lending.dart
import 'dart:convert';
import 'dart:typed_data';
import '../pubkey.dart';
import '../transaction.dart';

Uint8List _u64Le(BigInt n) {
  final buf = ByteData(8)..setUint64(0, n.toInt(), Endian.little);
  return buf.buffer.asUint8List();
}

Uint8List _u32Le(int n) {
  final buf = ByteData(4)..setUint32(0, n, Endian.little);
  return buf.buffer.asUint8List();
}

Uint8List _str(String s) {
  final bytes = Uint8List.fromList(utf8.encode(s));
  return Uint8List.fromList([..._u32Le(bytes.length), ...bytes]);
}

Uint8List _concat(List<Uint8List> a) {
  final t = a.fold(0, (s, x) => s + x.length);
  final o = Uint8List(t);
  int off = 0;
  for (final x in a) { o.setRange(off, off + x.length, x); off += x.length; }
  return o;
}

class LendingProgram {
  static HimshaInstruction initCollection(
    HimshaPublicKey collection, HimshaPublicKey payer, String name,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.lending,
        accounts: [
          AccountMeta.writable(collection, false),
          AccountMeta.writable(payer, true),
        ],
        data: _concat([Uint8List.fromList([0]), _str(name)]),
      );

  static HimshaInstruction placeBid(
    HimshaPublicKey collection, HimshaPublicKey lender,
    Uint8List bidTxid, int bidVout,
    BigInt loanValue, BigInt loanPeriod,
    String lenderOrdinalsAddr, String lenderPaymentsAddr,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.lending,
        accounts: [
          AccountMeta.writable(collection, false),
          AccountMeta.readonly(lender, true),
        ],
        data: _concat([
          Uint8List.fromList([1]),
          bidTxid, _u32Le(bidVout),
          _u64Le(loanValue), _u64Le(loanPeriod),
          _str(lenderOrdinalsAddr), _str(lenderPaymentsAddr),
        ]),
      );

  static HimshaInstruction acceptBid(
    HimshaPublicKey collection, HimshaPublicKey borrower,
    String inscriptionId, Uint8List inscriptionTxid, int inscriptionVout,
    String borrowerOrdinalsAddr, String borrowerPaymentsAddr,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.lending,
        accounts: [
          AccountMeta.writable(collection, false),
          AccountMeta.readonly(borrower, true),
        ],
        data: _concat([
          Uint8List.fromList([2]),
          _str(inscriptionId),
          inscriptionTxid, _u32Le(inscriptionVout),
          _str(borrowerOrdinalsAddr), _str(borrowerPaymentsAddr),
        ]),
      );

  static HimshaInstruction repayLoan(
    HimshaPublicKey collection, HimshaPublicKey borrower,
    String inscriptionId, Uint8List repayTxid, int repayVout,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.lending,
        accounts: [
          AccountMeta.writable(collection, false),
          AccountMeta.readonly(borrower, true),
        ],
        data: _concat([
          Uint8List.fromList([3]),
          _str(inscriptionId), repayTxid, _u32Le(repayVout),
        ]),
      );

  static HimshaInstruction claimDefault(
    HimshaPublicKey collection, HimshaPublicKey lender, String inscriptionId,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.lending,
        accounts: [
          AccountMeta.writable(collection, false),
          AccountMeta.readonly(lender, true),
        ],
        data: _concat([Uint8List.fromList([4]), _str(inscriptionId)]),
      );
}

File: himsha-sdk-python/himsha_sdk/programs/lending.py
"""HIMSHA Lending Program helpers."""

import struct
from ..pubkey import HimshaPublicKey, PROGRAM_IDS
from ..transaction import AccountMeta, HimshaInstruction


def _u64(n: int) -> bytes:
    return struct.pack("<Q", n)


def _u32(n: int) -> bytes:
    return struct.pack("<I", n)


def _str(s: str) -> bytes:
    encoded = s.encode()
    return _u32(len(encoded)) + encoded


def init_collection(
    collection: HimshaPublicKey, payer: HimshaPublicKey, name: str
) -> HimshaInstruction:
    return HimshaInstruction(
        program_id=PROGRAM_IDS.lending,
        accounts=[
            AccountMeta.writable(collection, False),
            AccountMeta.writable(payer, True),
        ],
        data=bytes([0]) + _str(name),
    )


def place_bid(
    collection: HimshaPublicKey,
    lender: HimshaPublicKey,
    bid_txid: bytes,        # 32 bytes
    bid_vout: int,
    loan_value_sats: int,
    loan_period_secs: int,
    lender_ordinals_addr: str,
    lender_payments_addr: str,
) -> HimshaInstruction:
    if len(bid_txid) != 32:
        raise ValueError("bid_txid must be 32 bytes")
    return HimshaInstruction(
        program_id=PROGRAM_IDS.lending,
        accounts=[
            AccountMeta.writable(collection, False),
            AccountMeta.readonly(lender, True),
        ],
        data=(
            bytes([1]) + bid_txid + _u32(bid_vout)
            + _u64(loan_value_sats) + _u64(loan_period_secs)
            + _str(lender_ordinals_addr) + _str(lender_payments_addr)
        ),
    )


def accept_bid(
    collection: HimshaPublicKey,
    borrower: HimshaPublicKey,
    inscription_id: str,
    inscription_txid: bytes,   # 32 bytes
    inscription_vout: int,
    borrower_ordinals_addr: str,
    borrower_payments_addr: str,
) -> HimshaInstruction:
    if len(inscription_txid) != 32:
        raise ValueError("inscription_txid must be 32 bytes")
    return HimshaInstruction(
        program_id=PROGRAM_IDS.lending,
        accounts=[
            AccountMeta.writable(collection, False),
            AccountMeta.readonly(borrower, True),
        ],
        data=(
            bytes([2]) + _str(inscription_id)
            + inscription_txid + _u32(inscription_vout)
            + _str(borrower_ordinals_addr) + _str(borrower_payments_addr)
        ),
    )


def repay_loan(
    collection: HimshaPublicKey,
    borrower: HimshaPublicKey,
    inscription_id: str,
    repay_txid: bytes,    # 32 bytes
    repay_vout: int,
) -> HimshaInstruction:
    if len(repay_txid) != 32:
        raise ValueError("repay_txid must be 32 bytes")
    return HimshaInstruction(
        program_id=PROGRAM_IDS.lending,
        accounts=[
            AccountMeta.writable(collection, False),
            AccountMeta.readonly(borrower, True),
        ],
        data=bytes([3]) + _str(inscription_id) + repay_txid + _u32(repay_vout),
    )


def claim_default(
    collection: HimshaPublicKey, lender: HimshaPublicKey, inscription_id: str
) -> HimshaInstruction:
    return HimshaInstruction(
        program_id=PROGRAM_IDS.lending,
        accounts=[
            AccountMeta.writable(collection, False),
            AccountMeta.readonly(lender, True),
        ],
        data=bytes([4]) + _str(inscription_id),
    )
